File: projects/mcp-servers/kotlin/src/main/kotlin/com/example/mcp/tools/GreetTool.kt
package com.example.mcp.tools

import io.modelcontextprotocol.kotlin.sdk.Tool
import io.modelcontextprotocol.kotlin.sdk.ToolHandler
import io.modelcontextprotocol.kotlin.sdk.ToolResult
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * A sample MCP tool that greets a user by name.
 *
 * Input schema:
 * - `name` (string, required): The name of the person to greet.
 *
 * Output:
 * - A JSON object containing a friendly greeting message.
 */
class GreetTool {

    /** Tool definition / specification exposed to the MCP client. */
    val definition: Tool = Tool(
        name = "greet",
        description = "Returns a personalised greeting for the given name",
        inputSchema = buildJsonObject {
            put("type", JsonPrimitive("object"))
            put("properties", buildJsonObject {
                put("name", buildJsonObject {
                    put("type", JsonPrimitive("string"))
                    put("description", JsonPrimitive("The name of the person to greet"))
                })
            })
            put("required", buildJsonObject {
                // The SDK expects required to be an array of strings
                // We pass it as a JSON array in the schema
            })
        }
    )

    /** Handler invoked when the client calls the `greet` tool. */
    val handler: ToolHandler = ToolHandler { arguments ->
        val name = arguments?.get("name")?.let { value ->
            if (value is JsonPrimitive && value.isString) {
                value.content
            } else {
                null
            }
        } ?: "World"

        val greeting = "Hello, $name! Welcome to the Kotlin MCP server."
        ToolResult(
            content = listOf(
                io.modelcontextprotocol.kotlin.sdk.Content(
                    type = "text",
                    text = greeting
                )
            )
        )
    }
}

File: projects/mcp-servers/ruby/lib/my_mcp_server/tools/greet_tool.rb
# frozen_string_literal: true

module MyMcpServer
  module Tools
    # A simple greeting tool.
    class GreetTool
      def self.name
        'greet'
      end

      def self.description
        'Greet a person by name with an optional title.'
      end

      def self.schema
        {
          type: 'object',
          properties: {
            name: {
              type: 'string',
              description: 'The name of the person to greet'
            },
            title: {
              type: 'string',
              description: 'Optional title (e.g. Mr., Ms., Dr.)'
            }
          },
          required: ['name']
        }
      end

      def self.execute(args)
        name = args['name']
        title = args['title']

        greeting = +'Hello'
        greeting << " #{title}" if title && !title.empty?
        greeting << ", #{name}!"

        { greeting: greeting }
      end
    end
  end
end
